intrahospital_api/update_demographics.py

```diff
@@ -220,13 +237,161 @@ def get_mrn_and_date_from_merge_comment(merge_comment):
     return sorted(result, key=lambda x: x[1], reverse=True)
 
 
+def get_all_merged_mrns_since(since):
+    query_result = api.execute_hospital_query(
+        GET_ALL_MERGED_MRNS_SINCE, params={"since": since}
+    )
+    return [i["Patient_Number"] for i in query_result]
+
+
 class MergeException(Exception):
     pass
 
 
 class CernerPatientNotFoundException(Exception):
     pass
 
+def mrn_in_elcid(mrn):
+    """
+    Returns True if the MRN is in Demographics.hospital_number
+    or MergedMRN.mrn fields
+    """
+    if models.MergedMRN.objects.filter(
+        mrn=mrn
+    ).exists():
+        return True
+    if models.Demographics.objects.filter(
+        hospital_number=mrn
+    ).exists():
+        return True
+    return False
+
+
+def check_and_handle_upstream_merges_for_mrns(mrns):
+    """
+    Takes in a list of MRNs.
+
+    Filters those not related to elCID.
+
+    If they are inactive, creates a Patient for the
+    active MRN and creates MergedMRN for all related inactive
+    MRNs.
+
+    If they are active, creates MergedMRN for all
+    related inactive MRNs.
+    """
+    mrn_to_upstream_merge_data = get_mrn_to_upstream_merge_data()
+    now = timezone.now()
+
+    all_active_mrn_and_merged_dicts = []
+    for mrn in mrns:
+        all_active_mrn_and_merged_dicts.append(get_active_mrn_and_merged_mrn_data(
+            mrn, mrn_to_upstream_merge_data
+        ))
+
+    unique_active_mrn_and_merged_dicts = []
+    # it is possible that the MRNs passed in will link to the same active MRN
+    # so make sure we only have one per active MRN
+    mrns_seen = set()
+    for active_mrn, merged_dicts in all_active_mrn_and_merged_dicts:
+        if active_mrn in mrns_seen:
+            continue
+        mrns_seen.add(active_mrn)
+        unique_active_mrn_and_merged_dicts.append((active_mrn, merged_dicts,))
+
+    # If neither the active MRN or the inactive MRN is in elcid then
+    # we do not need to process it.
+    active_mrn_and_merged_dicts = []
+    for active_mrn, merged_dicts in unique_active_mrn_and_merged_dicts:
+        if mrn_in_elcid(active_mrn):
+            active_mrn_and_merged_dicts.append((active_mrn, merged_dicts,))
+        elif len([i for i in merged_dicts if mrn_in_elcid(i["mrn"])]) > 0:
+            active_mrn_and_merged_dicts.append((active_mrn, merged_dicts,))
+    logger.info('Generating merged MRNs')
+    to_create = []
+    for active_mrn, merged_dicts in active_mrn_and_merged_dicts:
+        active_patient = Patient.objects.filter(
```

Comment:
May be missing something, but I think this spelling has you looping more times and setting up more datastructures than the way I would have spelled it...

```python
active_mrns_seen = set()
for mrn in mrns:
    active, merged_data = get_active_mrn_and_merged_mrn_data(mrn, mrn_to_upstream_merge_data)
    if active in active_mrns_seen:
        # The list passed into this function potentially contains all nodes in a merge graph
        continue 
    if not mrn_in_elcid(active):
        if not any(mrn_in_elcid(i['mrn'] for i in merged_data)):
            continue # no node in this graph is part of the elCID cohort
    active_patient = Patient.objects.filter(demographics__hospital_unmber=active).first()
```

Reply:
Sure, I thought multiple for loops made it more readable, because each did one thing. I'm happy with either.